include/oxen/quic/ip.hpp

```diff
@@ -160,48 +147,124 @@ namespace oxen::quic
             }
             return b;
         }
+
+        std::string to_string() const;
+        constexpr static bool to_string_formattable = true;
     };
 
-    struct ipv6_net
+    namespace detail
     {
-        ipv6 base;
-        uint8_t mask;
+        struct masked_ipv4
+        {
+            ipv4 ip;
+            uint8_t mask;
+
+          protected:
+            constexpr masked_ipv4() = default;
+            constexpr masked_ipv4(ipv4 b, uint8_t m) : ip{b}, mask{m} {}
+
+          public:
+            constexpr ipv4 max_ip() const
+            {
+                auto b = ip.to_base(mask);
 
-        constexpr ipv6 max_ip() const
+                if (mask < 32)
+                    b.addr |= (uint32_t{1} << (32 - mask)) - 1;
+
+                return b;
+            }
+
+            constexpr auto operator<=>(const masked_ipv4& other) const
+            {
+                return std::tie(ip, mask) <=> std::tie(other.ip, other.mask);
+            }
+
+            constexpr bool operator==(const masked_ipv4& other) const { return (*this <=> other) == 0; }
```

Comment:
I believe all of this can be replaced with just:
```suggestion
            constexpr auto operator<=>(const masked_ipv4&) const = default;
```
because if you have an explicitly defaulted <=> then == and != get implicitly defined in terms of it (unless otherwise specified).

And I think the same can be done for `ipv4`.

Reply:
On second though: having the <=> in the underlying base class introduces some weirdness we don't want (such as making `net < mask` a valid, if non-nonsensical, comparison), so probably better to just delete it from here completely and define separate comparators in the derived classes so that we don't allow that sort of comparison to compile.